ML Rule Suppression UI Improvements

rylnd opened the pull request:
## Summary

1. Re-enables and adds additional ML cypress tests

1. Adds ML fields to Define Step
    <img width="881" alt="Screenshot 2024-06-06 at 5 14 17 PM" src="https://github.com/rylnd/kibana/assets/657252/9aa6ed99-1e02-44a0-ad1b-785136510d68">

1. Disables suppression UI when no relevant ML jobs are enabled
    <img width="776" alt="Screenshot 2024-06-17 at 11 26 01 PM" src="https://github.com/rylnd/kibana/assets/657252/8adc590f-7cab-4429-8c8e-35fd3c04f9e8">


1. Adds warning text when _some_ relevant ML jobs are not enabled
    <img width="776" alt="Screenshot 2024-06-17 at 11 26 16 PM" src="https://github.com/rylnd/kibana/assets/657252/3e5eaf74-fa4f-47bb-a42d-3fe16303c92c">

rylnd committed:
Add copy for new ML rule warnings

rylnd committed:
Add UX messaging for ML suppression edge cases

* Disables suppression fields if no relevant ML jobs are running (as we
  cannot retrieve field info)
* Adds a warning message if not all relevant ML jobs are running (as we
  may be missing some field info)

Next step is testing this; we don't currently have a way to run ML rules
in cypress, but I'm going to attempt to copy the logic in our FTR to
accomplish this.

rylnd committed:
style: sort StepDefineRule arguments

rylnd committed:
Provide ML fields on Define step of rule creation

This was previously only available on the About step, via the
useRuleIndices hook in combination with the useFetchIndex hook.

Add new composite hook that encapsulates the same logic, and provides it
to the define step. Unlike on the About step, we are currently only
using this for ML fields as other situations derive their field list
from a passed prop (which might be a performance optimization, or a bug,
or both).

rylnd committed:
WIP: adding cypress tests for ML warning

To do this we need to get the ML API to recognize our jobs as installed
and running. They are currently _not_ recognizing this (although there
are anomalies in the index).

Still troubleshooting to see what's missing, here. This logic was
cribbed from the analogous FTR tests, but those also aren't working so
*shrug*.

rylnd committed:
Install test ML jobs with a known/compatible group

Specifying the `groups` parameter when using the "setup module" API
causes the corresponding jobs to be installed _with only the specified
group_. This meant that in our FTR tests, we have been installing jobs
with the `auditbeat` group.

However, part of the contract between ML and Detection Engine is that we
use the `group` parameter to determine relevance: if it doesn't belong
to either the `security` or (legacy) `siem` group(s), it effectively
does not exist to the Detection Engine.

This fixes the (very confusing) issue of jobs being installed but not
recognized, by specifying a recognized group id (and using our shared
constant for it), both in the FTR and cypress utilities.

rylnd committed:
Use correct job ID in cypress tests

I have seen the _ecs prefix in a few places, but I'm not quite sure if
it's actually part of official ML naming or not. Regardless, using the
incorrect name caused the "start datafeed" request to fail with a "no
datafeed for job ID" error.

rylnd committed:
Replace duplicated, magic-stringed function for existing one

The existing one also references the shared constants for our group IDs,
so 👍.

rylnd committed:
ML Suppression Cypress tests are green

* Cleans up debugging logs
* Adds helper for ensuring that jobs are not started at beginning of
  suite
* Fixes form filling utility to support single values for
  machine_learning_job_id
* Updates suppression fields now that we're actually using real fields
  from anomaly indices

yctercero commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/utils/machine_learning/machine_learning_setup.ts:
```diff
@@ -22,7 +23,7 @@ export const executeSetupModuleRequest = async ({
     .send({
       prefix: '',
-      groups: ['auditbeat'],
+      groups: [ML_GROUP_ID],
```
Amazing job sticking with it to figure this out. Woohoo to finally having our ML tests able to run.

nastasha-solomon commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/translations.tsx:
```diff
@@ -234,6 +234,21 @@ export const EQL_SEQUENCE_SUPPRESSION_GROUPBY_VALIDATION_TEXT = i18n.translate(
+export const MACHINE_LEARNING_SUPPRESSION_DISABLED_LABEL = i18n.translate(
+  'xpack.securitySolution.detectionEngine.createRule.stepDefineRule.machineLearningSuppressionDisabledLabel',
+  {
+    defaultMessage: 'Machine Learning jobs must be running to enable alert suppression.',
```
While this is technically true, I think we can still be more direct and tell users that they must act to enable suppression. 

```suggestion
    defaultMessage: 'To enable alert suppression, start one or more Machine Learning jobs.',
```

Another, more general option: 
To enable alert suppression, start Machine Learning jobs.

nastasha-solomon commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/translations.tsx:
```diff
@@ -234,6 +234,21 @@ export const EQL_SEQUENCE_SUPPRESSION_GROUPBY_VALIDATION_TEXT = i18n.translate(
+  'xpack.securitySolution.detectionEngine.createRule.stepDefineRule.machineLearningSuppressionIncompleteLabel',
+  {
+    defaultMessage:
+      'This list of fields may be incomplete as some Machine Learning jobs are not running.',
```
What should users do if they notice that certain fields are missing? We should prob include that suggestion so they have a path forward from here.

rylnd replied:
We could provide the list of non-running ML jobs, but that could get verbose. As with the above, the solution is to start ML jobs.

nastasha-solomon replied:
Makes sense. I also wonder how useful that list would be for users who don't know which jobs to enable. 

How's this:

```suggestion
      'This list of fields might be incomplete as some Machine Learning jobs are not running. Start more jobs to add more fields.',
```

rylnd committed:
More direct copy for user action

Co-authored-by: Nastasha Solomon <79124755+nastasha-solomon@users.noreply.github.com>

rylnd committed:
Replace display names with actual Job IDs

The way the cypress helper that consumes these is written, both of these
forms work, but we're never going to encounter a rule with the display
name params, and knowing the display name but not the ID is not useful
for investigative purposes.

I can see how this might have been done to prevent needing to change
these jobs as their IDs change, but I think it's more likely that those
will change than their IDs.

rylnd committed:
Extract ML rule validation logic into hook

There's not a lot here, but I feel bad for adding anything to
step_define_rule so this is an attempt to minimize that.

In the course of refactoring I also caught a bug (perhaps just a test
environment one) where the form fields are temporarily `undefined` when
the hooks are run. I updated the form type to reflect this; hopefully
that doesn't have broader impact (but if it does, those are probably
also uncaught bugs).

rylnd committed:
Relax requirement on number of rules

The combination of shared state and retry logic means that asserting
exactly 1 rule exists will never work if rule creation succeeds in a
previous step. If we instead assert that there is _at least_ the
expected number of rules, we have a chance of the retry working.

rylnd committed:
Clean up existing ML cypress test

* Stop datafeeds before creating rule
* Simplify jobId logic

rylnd committed:
Bring back job display names for assertions

Turns out the reason the "Job IDs" were persisted as human-readable text
was so that they could be reused for assertions.

I still think these should be separate, so I'm adding them back for this
specific assertion.

rylnd committed:
Re-enable ML rule edit tests

* Adds necessary setup/teardown for ML integration

rylnd committed:
Robustify ML cypress tasks

1. Use "proper" combobox text, and capture it within a helper method
I swear I saw this working when I was doing the same stuff for the ML
Job picker, but I must have only been dealing with one item, or the
items I was selecting were somehow different. Downarrow is _required_ on
the first option (a simple "enter" will select nothing), but using
downarrow on subsequent options will cause the _second_ suggested item
to be selected. E.g. if I type "by_field_value", it suggests both
"by_field_value" and "client.by_field_value," and {downarrow}{enter}
would cause the latter to be selected.

2. I also ensure that our new ML validations have run (which causes
  suppression fields to be disabled) before attempting to interact with
  the suppression fields, as this was causing some flakiness now that
  these checks are done async

3. I also fixed the broken `clearAlertSuppressionFields` task, which had
  never work but also had never been exercised since the relevant test
  was skipped.

rylnd committed:
Ensure test has clean setup

There were no less than four assertions in this test that relied on
there being no other rules present in the environment, but nothing was
being done to ensure that was the case. I can't imagine why these were
skipped!

rylnd committed:
Update ML warning copy per docs' suggestions

Co-authored-by: Nastasha Solomon <79124755+nastasha-solomon@users.noreply.github.com>

rylnd committed:
Remove exclusivity from FTR tests, add TODO for re-skipping

I want to run these in the flaky runner to get a sense of how/where
they're still failing, for now.

rylnd committed:
Fix suppression fields for non-ML cases

We were over-eagerly disabling these fields when the ML checks were not
relevant.
